csrc/scheduler/greedy.cpp

```diff
@@ -0,0 +1,730 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2025-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+
+#include <debug.h>
+#include <device_lower/analysis/fusion_info.h>
+#include <device_lower/analysis/sync_information.h>
+#include <device_lower/utils.h>
+#include <exceptions.h>
+#include <id_model/id_model.h>
+#include <id_model/to_string.h>
+#include <instrumentation.h>
+#include <ir/utils.h>
+#include <iter_visitor.h>
+#include <options.h>
+#include <scheduler/debug_utils.h>
+#include <scheduler/greedy.h>
+#include <scheduler/runtime_info.h>
+#include <scheduler/tools/maxinfo_propagator.h>
+#include <scheduler/utils.h>
+#include <transform_replay.h>
+
+#include <ATen/cuda/CUDAContext.h>
+
+#include <ranges>
+#include <vector>
+
+namespace nvfuser {
+
+namespace {
+
+// These are the current supported constrained ops.
+std::vector<Expr*> getAllConstrainedOps(Fusion* fusion) {
+  return ir_utils::getOpsOfType<ArgsortOp, ScanOp, PadOp>(fusion);
+}
+
+class CompileTimeChecker : private IterVisitor {
+ public:
+  static bool run(Fusion* fusion, const ValGraph& exact_graph) {
+    CompileTimeChecker checker(fusion, exact_graph);
+    if (!checker.can_schedule_ && !checker.reject_reason_.empty()) {
+      scheduler_debug_utils::canScheduleRejectReason(
+          SchedulerType::Greedy, checker.reject_reason_);
+    }
+    return checker.can_schedule_;
+  }
+
+ private:
+  CompileTimeChecker(Fusion* fusion, const ValGraph& exact_graph)
+      : exact_graph_(exact_graph) {
+    traverse(fusion);
+    // If this fusion requires the exact block dimension, requires the
+    // constrained IDs to be exactly mapped. This is not necessary but
+    // sufficient.
+    if (needs_exact_block_dim_ && mismatched_constrained_id_detected_) {
+      can_schedule_ = false;
```

Comment:
nitpick, since we are being conservative here, I think it's even more important to give a rejection reason.

Hint, we can move the code comment into reject_reason_.